drivers/pwm/pwm_nrfx.c

```diff
@@ -284,63 +290,71 @@ static int pwm_nrfx_init(const struct device *dev)
 		seq_values_ptr_get(dev)[i] = PWM_NRFX_CH_VALUE(0, inverted);
 	}
 
-	nrfx_err_t result = nrfx_pwm_init(&config->pwm,
-					  &config->initial_config,
-					  NULL,
-					  NULL);
-	if (result != NRFX_SUCCESS) {
-		LOG_ERR("Failed to initialize device: %s", dev->name);
-		return -EBUSY;
-	}
-
 	return 0;
 }
 
-#ifdef CONFIG_PM_DEVICE
-static void pwm_nrfx_uninit(const struct device *dev)
+static int pwm_suspend(const struct device *dev)
 {
 	const struct pwm_nrfx_config *config = dev->config;
+	struct pwm_nrfx_data *data = dev->data;
 
-	nrfx_pwm_uninit(&config->pwm);
+	if (data->active) {
+		/* PWM is still in use. Force stopping. */
+		data->stop_requested = true;
+		nrfx_pwm_stop(&config->pwm, false);
+		while (data->stop_requested) {
+		}
+	}
 
 	memset(dev->data, 0, sizeof(struct pwm_nrfx_data));
+
+	if (IS_ENABLED(CONFIG_PM_DEVICE)) {
+		return pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
+	}
+
+	return 0;
 }
 
 static int pwm_nrfx_pm_action(const struct device *dev,
 			      enum pm_device_action action)
 {
-	const struct pwm_nrfx_config *config = dev->config;
-	int ret = 0;
+	int ret = -ENOTSUP;
 
 	switch (action) {
-	case PM_DEVICE_ACTION_RESUME:
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
-		ret = pwm_nrfx_init(dev);
-		break;
-
 	case PM_DEVICE_ACTION_SUSPEND:
-		pwm_nrfx_uninit(dev);
-
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
-		}
+		ret = pwm_suspend(dev);
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		ret = pwm_resume(dev);
 		break;
-
 	default:
-		return -ENOTSUP;
+		break;
 	}
 
 	return ret;
 }
-#else
 
-#define pwm_nrfx_pm_action NULL
+static int pwm_nrfx_init(const struct device *dev)
+{
+	const struct pwm_nrfx_config *config = dev->config;
+	nrfx_err_t err;
+	int ret;
+
+	ANOMALY_109_EGU_IRQ_CONNECT(NRFX_PWM_NRF52_ANOMALY_109_EGU_INSTANCE);
 
-#endif /* CONFIG_PM_DEVICE */
+	err = nrfx_pwm_init(&config->pwm, &config->initial_config, pwm_handler, dev->data);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to initialize device: %s", dev->name);
+		return -EBUSY;
+	}
+
+	ret = pwm_suspend(dev);
```

Comment:
Why not just call `pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);` here (or better before the nrfx driver is initialized, for consistency with other shims)? That's essentially what is needed, right?